source/h2_stream.c

```diff
@@ -211,22 +225,141 @@ struct aws_h2_stream *aws_h2_stream_new_request(
     /* Init H2 specific stuff */
     stream->thread_data.state = AWS_H2_STREAM_STATE_IDLE;
     stream->thread_data.outgoing_message = options->request;
-    aws_http_message_acquire(stream->thread_data.outgoing_message);
 
+    aws_linked_list_init(&stream->synced_data.pending_frame_list);
+    if (aws_mutex_init(&stream->synced_data.lock)) {
+        AWS_H2_STREAM_LOGF(
+            ERROR, stream, "Mutex init error %d (%s).", aws_last_error(), aws_error_name(aws_last_error()));
+        aws_mem_release(stream->base.alloc, stream);
+        return NULL;
+    }
+    aws_http_message_acquire(stream->thread_data.outgoing_message);
+    aws_channel_task_init(
+        &stream->cross_thread_work_task, s_stream_cross_thread_work_task, stream, "HTTP/2 stream cross-thread work");
     return stream;
 }
 
+static void s_stream_cross_thread_work_task(struct aws_channel_task *task, void *arg, enum aws_task_status status) {
+    (void)task;
+    if (status != AWS_TASK_STATUS_RUN_READY) {
+        return;
+    }
+
+    struct aws_h2_stream *stream = arg;
+    struct aws_h2_connection *connection = s_get_h2_connection(stream);
+
+    struct aws_linked_list pending_frames;
+    aws_linked_list_init(&pending_frames);
+
+    size_t window_update_size;
+
+    { /* BEGIN CRITICAL SECTION */
+        s_lock_synced_data(stream);
+        stream->synced_data.is_cross_thread_work_task_scheduled = false;
+
+        aws_linked_list_swap_contents(&stream->synced_data.pending_frame_list, &pending_frames);
+        /* window_update_size is ensured to be not greater than AWS_H2_WINDOW_UPDATE_MAX */
+        window_update_size = stream->synced_data.window_update_size;
+        stream->synced_data.window_update_size = 0;
+
+        s_unlock_synced_data(stream);
+    } /* END CRITICAL SECTION */
+
+    while (!aws_linked_list_empty(&pending_frames)) {
+        struct aws_linked_list_node *node = aws_linked_list_pop_front(&pending_frames);
+        struct aws_h2_frame *frame = AWS_CONTAINER_OF(node, struct aws_h2_frame, node);
+        aws_h2_connection_enqueue_outgoing_frame(connection, frame);
+    }
+    /* We already enqueued the window_update frame, just apply the change and let our peer check this value. No matter
+     * overflow happens or not, peer will dectect it for us. */
+    if (aws_h2err_failed(aws_h2_stream_window_size_change(stream, (int32_t)window_update_size, true /*self*/))) {
```

Comment:
It's possible that the peer does NOT detect this.
If the peer sends some DATA before this WINDOW_UPDATE is received, then it might **not** cause overflow on the peer's side, and then the 2 sides are just out of sync.

Probably best to just kill the connection if we overflow locally.
Detecting the overflow from within the public update_window() fn would require that we store the current window in synced_data, which seems like an unnecessary amount of synchronization.
We could add a new AWS_ERROR like AWS_ERROR_HTTP_FLOW_CONTROL_WINDOW_OVERFLOW (we'd send AWS_H2_INTERNAL_ERROR to the peer)

Reply:
you are right, but let's take another approach ensures the sync between remote and local to address this problem.